lib/ProductOpener/Auth.pm

```diff
@@ -0,0 +1,743 @@
+		($refresh_token, $refresh_expires_at, $access_token, $access_expires_at) = refresh_access_token($refresh_token);
+		unless ($access_token) {
+			start_authorize($request_ref);
+			return;
+		}
+	}
+
+	# ID Token validation
+	#my $id_token = OIDC::Lite::Model::IDToken->load($token->id_token);
+
+	$log->info('request is ok', $request_ref) if $log->is_info();
+
+	return;
+}
+
+=head2 start_signout($request_ref)
+
+Initiates the sign-out process by redirecting the user to the authorization page.
+
+=head3 Arguments
+
+=head4 A reference to a hash containing request information. $request_ref
+.
+=head3 Return Values
+
+None
+=cut
+
+sub start_signout ($request_ref) {
+	my $return_url = single_param('return_url');
+	if (   (not $return_url)
+		or (not($return_url =~ /^https?:\/\/$subdomain\.$server_domain/)))
+	{
+		$return_url = $formatted_subdomain;
+	}
+
+	my $id_token = $request_ref->{id_token};
+	unless ($User_id and $id_token) {
+		param('length', 'logout');
+		init_user($request_ref);
+		redirect_to_url($request_ref, 302, $formatted_subdomain);
+		return;
+	}
+
+	_ensure_oidc_is_discovered();
+
+	my $nonce = generate_token(64);
+	my $end_session_endpoint = $oidc_discover_document->{end_session_endpoint};
+	my $redirect_url
+		= $end_session_endpoint
+		. '?post_logout_redirect_uri='
+		. uri_escape($signout_callback_uri)
+		. '&id_token_hint='
+		. uri_escape($id_token)
+		. '&state='
+		. uri_escape($nonce);
+
+	$request_ref->{cookie} = generate_oidc_cookie($nonce, $return_url);
+	redirect_to_url($request_ref, 302, $redirect_url);
+	return;
+}
+
+=head2 signout_callback($request_ref)
+
+Handles the callback after successful sign-out, clears session cookie.
+
+=head3 Arguments
+
+=head4 A reference to a hash containing request information. $request_ref
+
+=head3 Return values
+
+The return URL after successful sign-out.
+=cut
+
+sub signout_callback ($request_ref) {
+	if (not(defined cookie($cookie_name))) {
+		return $formatted_subdomain;
+	}
+
+	my $state = single_param('state');
+	my %cookie_ref = cookie($cookie_name);
+	my $nonce = $cookie_ref{'nonce'};
+	if (not($state eq $nonce)) {
+		$log->info('unexpected nonce', {nonce => $nonce, expected_nonce => $state}) if $log->is_info();
+		display_error_and_exit('Invalid Nonce during OIDC logout', 500);
+	}
+
+	param('length', 'logout');
+	init_user($request_ref);
+
+	return $cookie_ref{'return_url'};
+}
+
+sub create_user_in_keycloak ($user_ref, $password) {
+	my $keycloak_users_endpoint = $oidc_options{keycloak_users_endpoint};
+	unless ($keycloak_users_endpoint) {
+		display_error_and_exit('keycloak_users_endpoint not configured', 500);
+	}
+
+	my $token = get_token_using_client_credentials();
+	unless ($token) {
+		display_error_and_exit('Could not get token to manage users with keycloak_users_endpoint', 500);
+	}
+
+	my $api_request_ref = {
+		email => $user_ref->{email},
+		emailVerified => $JSON::PP::true,    # TODO: Keep this for compat with current register endpoint?
```

Comment:
We briefly touched this topic on Slack. One issue with email verification is that this breaks the "create user" API, which is currently used by some apps/SDKs. [Test case here](https://github.com/openfoodfacts/openfoodfacts-server/blob/9db5cd4c616b457681cea8b9567fa24a9459fa98/tests/integration/create_user.t). Currently, the assumption after successfully posting to `cgi/user.pl` is to get an active user that can be used to sign in and edit products directly. Enabling email verification would break with that idea. As far as I know, we don't have a way to communicate that changed behaviour to end-users right now.

Reply:
You are right. 

Could we, none the less, add a parameter to ask for email validation ? (so default would be no validation, but we could use validation, eg. on the web).
Do it only if it's easy to do.

Reply:
Parameter added. Defaulted to false for now.